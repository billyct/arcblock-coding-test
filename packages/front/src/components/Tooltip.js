import {useRef, useState, useEffect} from 'react'
import PropTypes from 'prop-types'
import cn from 'classnames'

const Tooltip = ({children, content, show}) => {
  const childrenRef = useRef()
  const tooltipRef = useRef()

  const [tooltipStyle, setTooltipStyle] = useState({})

  // content 变化之后，需要重新计算
  useEffect(() => {
    // 计算 tooltip 的 offset 样式
    const topOffset = -tooltipRef.current.clientHeight - 2
    const leftOffset = -tooltipRef.current.clientWidth / 2 + childrenRef.current.clientWidth / 2

    setTooltipStyle({
      top: topOffset,
      left: leftOffset,
    })

  }, [content])

  return (
    <div className='relative'>
      <div className='inline-block' ref={childrenRef}>
        {children}
      </div>
      <div
        ref={tooltipRef}
        data-testid='Tooltip__tooltip'
        className={cn(
          'bg-black opacity-80 text-white px-2 py-1 rounded text-xs leading-none absolute whitespace-nowrap',
          {
            'invisible': !show,
            'visible': show,
          },
        )}
        style={tooltipStyle}
      >
        {content}
        <div
          className='w-2 h-2 transform rotate-45 absolute bg-black left-0 right-0 mx-auto'
          style={{
            zIndex: -1,
            bottom: -2,
          }}
        />
      </div>
    </div>
  )
}

Tooltip.propTypes = {
  show: PropTypes.bool,
  content: PropTypes.string.isRequired,
}

Tooltip.defaultProps = {
  show: false,
}

export default Tooltip
