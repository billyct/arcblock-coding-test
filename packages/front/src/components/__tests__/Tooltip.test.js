import {render} from '@testing-library/react'
import Tooltip from '../Tooltip'

describe('test Tooltip Component', () => {
  test('snapshot should correctly',() => {
    const {container} = render(<Tooltip content={'a content'}/>)
    expect(container).toMatchSnapshot()
  })

  describe.each([
    [false, 'invisible'],
    [true, 'visible'],
  ])('test show attribute case %#', (show, classname) => {
    it(`should have class ${classname}`, function () {
      const {getByTestId} = render(<Tooltip content={''} show={show}/>)
      expect(getByTestId('Tooltip__tooltip')).toHaveClass(classname)
    })
  })
})
