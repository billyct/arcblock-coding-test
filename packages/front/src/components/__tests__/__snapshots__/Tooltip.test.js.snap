// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Tooltip Component snapshot should correctly 1`] = `
<div>
  <div
    class="relative"
  >
    <div
      class="inline-block"
    />
    <div
      class="bg-black opacity-80 text-white px-2 py-1 rounded text-xs leading-none absolute whitespace-nowrap invisible"
      data-testid="Tooltip__tooltip"
      style="top: -2px; left: 0px;"
    >
      a content
      <div
        class="w-2 h-2 transform rotate-45 absolute bg-black left-0 right-0 mx-auto"
        style="z-index: -1; bottom: -2px;"
      />
    </div>
  </div>
</div>
`;
